File: scripts/make_plots.py
import matplotlib.pyplot as plt
import seaborn as sns
import polars as pl
import numpy as np
from numpy import unique

results = pl.read_csv("output/results_test.csv")
groups = unique(results['group'])
plot_cols = ["#156082", "#3b7d23", "#78206e"]
sns.set_context("notebook", font_scale=1.5)

#### Visualization of one simulation ####
# Filter the data for replicate == 0 and initial_coverage_scenario == "low"
replicate_0_results = results.filter((pl.col("replicate") == 0) & (pl.col("initial_coverage_scenario") == "optimistic") & (pl.col("connectivity_scenario") == 1.0))

# Plot the line plot for each group
plt.figure()

for group, color in zip(groups, plot_cols):
    group_data = replicate_0_results.filter(pl.col("group") == group)
    sum_I1_I2 = group_data["I1"] + group_data["I2"]
    plt.plot(group_data["t"], sum_I1_I2, label=f'Group {group}', color=color)

plt.xlabel('Time (Days)')
plt.ylabel('Infected Population')
plt.legend()
plt.tight_layout()

plt.savefig('output/infection_curve.png')
plt.close()

#### Faceted Histogram Sub Groups ####
vax_scenario_order = ["low", "medium","optimistic"]

filtered_results = results.filter(pl.col("t") == 365)

# Convert to pandas DataFrame for seaborn compatibility
filtered_results_df = filtered_results.to_pandas()

# Calculate the bin edges for consistent bin width
min_value = filtered_results_df["Y"].min()
max_value = filtered_results_df["Y"].max()
bins = np.linspace(min_value, max_value, 100)  # 40 bins

# Create the faceted histogram with the same bin width for all histograms
g = sns.FacetGrid(filtered_results_df,
                  row="initial_coverage_scenario",
                  col = "connectivity_scenario",
                  row_order=vax_scenario_order,
                  sharex=True, sharey=True, height=4, aspect=3)
g.map_dataframe(sns.histplot, x="Y", hue="group", multiple="stack", bins=bins, palette=plot_cols)
# To set the x-axis to log scale: g.set(xscale="log")

g.set_axis_labels('Final size', 'Number of simulations')
g.add_legend(title='Group')
g.tight_layout()

g.savefig('output/final_size_groups.png')

#### Faceted Histogram Total outbreak size ####
# Filter the data for t=365

filtered_results = results.filter(pl.col("t") == 365)

# Convert to pandas DataFrame for seaborn compatibility
filtered_results_df = filtered_results.to_pandas()

# Group by simulation, vaccination scenario, and connectivity scenario, and sum the Y column
grouped_results = filtered_results_df.groupby(['replicate', 'initial_coverage_scenario', 'connectivity_scenario']).agg({'Y': 'sum'}).reset_index()
grouped_results.rename(columns={'Y': 'total_outbreak_size'}, inplace=True)

# Calculate the bin edges for consistent bin width
min_value = grouped_results["total_outbreak_size"].min()
max_value = grouped_results["total_outbreak_size"].max()
bins = np.linspace(min_value, max_value, 100)  # 20 bins

# Create the faceted histogram with the same bin width for all histograms
g = sns.FacetGrid(grouped_results,
                  row="initial_coverage_scenario",
                  col="connectivity_scenario",
                  row_order=vax_scenario_order,
                  sharex=True, sharey=True, height=4, aspect=3)
g.map_dataframe(sns.histplot, x="total_outbreak_size", bins=bins)
g.set_axis_labels('Total Outbreak Size', 'Number of Simulations')

plt.tight_layout()
plt.savefig('output/final_size_total.png')
plt.close()

File: scripts/make_plots.R
library(tidyverse)
library(ggplot2)

results <- read_csv("output/results_test.csv")
reps <- max(results$replicate)

plot_cols <- c("#20419a", "#cf4828", "#f78f47")

#### Cumulative and incidence plots ####
vax_levs <- c("low", "medium", "optimistic")
sub_conns <- c(0.01, 0.1)
filtered_results <- results |>
    filter(
        replicate %in% 1:20,
        initial_coverage_scenario %in% vax_levs,
        k_21 %in% sub_conns
    )

p <- filtered_results |>
    ggplot() +
    geom_line(
        aes(t, Y,
            col = factor(group),
            group = interaction(replicate, group)
        ),
        alpha = 0.5
    ) +
    facet_grid(k_21 ~ initial_coverage_scenario,
        labeller = label_both
    ) +
    theme_minimal(base_size = 18) +
    scale_color_manual(values = plot_cols) +
    # facet_wrap(~replicate) +
    labs(x = "Days", y = "Cumulative Infections", col = "Group")

ggsave(filename = paste0(
        "output/cumulative_curves_r0",
        12, ".png"
    ), plot = p, width = 10, height = 8)

#### Incidence plot ####
p <- filtered_results |>
    ggplot(aes(t, I1 + I2,
    col = factor(group),
    group=interaction(replicate, group))) +
    geom_line(alpha = 0.5) +
    facet_grid(k_21 ~ initial_coverage_scenario,
        labeller = label_both
    ) +
    theme_minimal(base_size = 18) +
    scale_color_manual(values = plot_cols) +
    labs(x = "Days", y = "Incident Infections", col = "Group")
ggsave(filename = paste0(
        "output/incidence_curves_r0",
        12, ".png"
    ), plot = p, width = 10, height = 8)


#### Overeall final size plot
for (i in c(12)) {
    p <- results |>
        filter(
            t == 365,
            initial_coverage_scenario %in% vax_levs,
            k_21 %in% sub_conns
        ) |>
        group_by(replicate, initial_coverage_scenario, k_21) |>
        summarise(final_size = sum(Y)) |> # total sum across groups
        ggplot(aes(final_size)) +
        # scale_x_log10() +
        geom_histogram(bins = 50) +
        theme_minimal(base_size = 18) +
        labs(x = "Final Outbreak Size", y = "Number of Simulations") +
        facet_grid(k_21 ~ initial_coverage_scenario,
            labeller = label_both)

    ggsave(filename = paste0(
        "output/overall_final_size_r0",
        i, ".png"
    ), plot = p, width = 10, height = 8)
}


#### Grouped outbreak size plots ####
for (i in c(12)) {
    p <- results |>
        filter(
            t == 365, k_21 %in% sub_conns,
            initial_coverage_scenario %in% vax_levs
        ) |>
        ggplot(aes(Y + 1, fill = factor(group))) +
        geom_histogram(position = "identity", alpha = 0.5) +
        scale_x_log10() +
        facet_grid(k_21 ~ initial_coverage_scenario,
            labeller = label_both
        ) +
        scale_fill_manual(values = plot_cols) +
        theme_minimal(base_size = 18) +
        labs(
            title = "R0=12", x = "Final Group Outbreak Size",
            y = "Number of simulations", fill = "Group"
        )
    ggsave(filename = paste0(
        "output/group_final_size_r0",
        i, ".png"
    ), plot = p, width = 10, height = 8)
}

#### Summary table
outbreak_sizes <- c(300, 500)
for (i in outbreak_sizes) {
    res_table <- results |>
        filter(t == 365, Y >= i) |>
        group_by(
            group,
            initial_coverage_scenario,
            k_21
        ) |>
        count() |>
        mutate(n = round(n / reps, 2) * 100) |>
        pivot_wider(names_from = group, values_from = n) |>
        select(
            InitialCoverage = initial_coverage_scenario,
            SubPopConnectivity = k_21,
            Sub1 = `1`, Sub2 = `2`,
            General = `0`
        )

    write_csv(res_table,
              paste0("output/res_table_outbreak_size", i, ".csv"))
}
